---
title: "Analyze SNV Data"
author: "Fong Chun Chan (fongchunchan@gmail.com)"
date: "May 22, 2015"
output: html_document
---

# Introduction
---

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

Inside of RStudio when you click the **Knit** button in the document, it will be generate an html includes both content as well as the output of any embedded R code chunks within the document. 

We will be using this R Markdown document to explore the SNV data in more detail. First let's load some packages that are needed

```{r setup}
library("data.table")
library("ggplot2")
library("plyr")
library("dplyr")
library("stringr")
```

# Strelka Analysis
---

Now let's load the Strelka results:

```{r}
strelkaDt <- fread("~/cbw-2015/snvs/content/passed.somatic.snvs.txt")

# rename columns so easier to reference later on
setnames(strelkaDt, "#CHROM", "CHROM")
setnames(strelkaDt, colnames(strelkaDt), gsub("GEN\\[0\\]", "normal", colnames(strelkaDt)))
setnames(strelkaDt, colnames(strelkaDt), gsub("GEN\\[1\\]", "tumour", colnames(strelkaDt)))

# convert CHROM variable in an ordered factor
strelkaDt[, CHROM := factor(CHROM, levels = c(1:22, "X"))]

# Converting QSS scores to probability
strelkaDt[, QSS_prob := round(1-10^(-QSS_NT/10), 3)]

# Generating allelic ratios 
strelkaDt[ALT == "A", tumour.AR := as.numeric(unlist(str_split(strelkaDt[1, tumour.AU], ","))[1]) / tumour.DP]
strelkaDt[ALT == "C", tumour.AR := as.numeric(unlist(str_split(strelkaDt[1, tumour.CU], ","))[1]) / tumour.DP]
strelkaDt[ALT == "G", tumour.AR := as.numeric(unlist(str_split(strelkaDt[1, tumour.GU], ","))[1]) / tumour.DP]
strelkaDt[ALT == "T", tumour.AR := as.numeric(unlist(str_split(strelkaDt[1, tumour.TU], ","))[1]) / tumour.DP]

strelkaDt[ALT == "C", tumour.AR := as.numeric(unlist(str_split(strelkaDt[1, tumour.CU], ","))[1]) / tumour.DP]

strelkaDt
strelkaDt
          
          

kstr_c("tumour.", strelkaDt[, ALT], "U")
strelkaDt %>%
  select(one_of(c("test_DP", "tumour.GU")))

strelkaDt

for (base in bases) {

  print(paste("Calculating the variant ratio for the", base, "variant..."))
_
  strelkaDt.baseFreq <- 
    strelkaDt.baseFreq %>%
    select(one_of(c("test_DP", str_c("test_", base, "U")))) %>% 
    mutate(test1 = 
    separate_(str_c("test_", base, "U"), into = c("test1", "test2"), sep = ",") %>%  # the value is comma-separated (tier1,tier2)
    mutate(testFreq = as.numeric(test1) / test_DP) %>%  # only consider tier1 reads in the variant ratio calculation
    plyr::rename( c("testFreq" = str_c("test_", base, "U_FREQ"))) %>%
    select(one_of(str_c("test_", base, "U_FREQ"))) %>%
    bind_cols(strelkaDt.baseFreq, .)
```

Let's look at the mutational load (i.e. number of mutations) across the genome.

```{r}
strelkaDt %>%
  filter(CHROM %in% c(1:22, "X")) %>%
  ggplot(aes(x = CHROM)) +
  geom_bar() +
  xlab("Chromosome") + 
  ylab("Number of Mutations") +
  ggtitle("Mutational Load")
```

```{r}

```


Let's look at some of the substitution patterns:

```{r}
strelkaDt %>%
  mutate(subPattern = str_c(REF, "->", ALT)) %>%
  ggplot(aes(x = subPattern)) +
  geom_bar()
```
